### PR TITLE
Prepare to switch from QSharedPointer to std::shared_ptr

### PR DESCRIPTION
**Pull request by uklotzde**
Some refactorings around the usage of Track/TrackPointer that lay the ground for replacing QSharedPointer with std::shared_ptr eventually.

The design of std::shared_ptr allows more efficient allocation of reference-counted objects by allocating the managing count object together with the managed object in a single memory allocation on the heap. But even more important: It's the standard! We already propose to use std::unqiue_ptr instead of QScopedPointer since upgrading to C++11.

The remaining changes to actually perform this switch are very simple. I've already prepared another branch for demonstration purposes: https://github.com/uklotzde/mixxx/tree/std_shared_ptr

### CODE REVIEW

File: src/engine/bpmcontrol.cpp [hunk 1]
```diff
@@ -739,7 +739,7 @@ void BpmControl::slotAdjustRateSlider() {
 void BpmControl::trackLoaded(TrackPointer pNewTrack, TrackPointer pOldTrack) {
     Q_UNUSED(pOldTrack);
     if (m_pTrack) {
-        disconnect(m_pTrack.data(), SIGNAL(beatsUpdated()),
+        disconnect(&*m_pTrack, SIGNAL(beatsUpdated()),
```
Comment (daschuer) [inline src/engine/bpmcontrol.cpp]: Note, If we check always for nullptr, we have no issue, but there would be a difference if not.
If m_pTrack is nullptr, get() will return nullptr while the behavior of &* is undefined. However it is practical nullptr according to this: 
http://stackoverflow.com/questions/2727834/c-standard-dereferencing-null-pointer-to-get-a-reference

File: src/track/track.cpp [hunk 2]
```diff
@@ -25,8 +25,17 @@ SecurityTokenPointer openSecurityToken(
     }
 }
 
+struct TrackDeleter {
```
Comment (daschuer) [inline src/track/track.cpp]: Why we need a functor here, and not just? 

void trackDeleter(Track* pTrack) {
       if (pTrack != nullptr) {
           pTrack->deleteLater();
       }
 }

**Review by daschuer — COMMENTED** [review]
Thank you for the PR. 
I have left only some questions.

**Comment by rryan** [comment]
I'm not opposed, but one benefit of QSharedPointer is that it has some QObject-specific handling that emits warnings if you e.g. delete a QObject that's managed by a QSharedPointer (it prevents a crash and prints a warning by clearing the shared pointer). This can happen unintentionally in Qt via the parent/child tree.

**Comment by daschuer** [comment]
Thank you for the reminder. This confirms that we should in general avoid shared pointers when the Qt object tree is involved. 

In this case the std version has some benefits. But what does it mean for a rule of thumb for Mixxx? 

A conceptual issue here, are the various connect() calls, since a plain pointer is stored elsewhere, which is normally an anti pattern when using shared pointers. In this case it is OK because of the delete later though. 
We could encapsulate the connect calls ... and add more boilerplate code ...

**Comment by uklotzde** [comment]
Great feedback, guys :+1:

Hopefully I found some nice Qt compliant solutions by introducing derived classes instead of using typedefs for TrackPointer and CuePointer. It is no longer possible to create shared pointers that violate the Qt rules for QObject deletion. The usage of CuePointer in the existing code was already wrong by using the default operator delete!

**Comment by daschuer** [comment]
Nice! Did you consider to move get() into the TrackPointer class? I am honestly unsure if typing saving &* or explicit get() is the way to go.

**Comment by uklotzde** [comment]
Yes I did, but I forgot to add it. Maybe tomorrow ;) I'm still undecided. 
Currently the derived classes only provide a limited set of constructors and 
don't modify the interface of their base classes in any way.

**Comment by uklotzde** [comment]
Temporarily adding get() to the derived (smart) pointer classes proved to be 
a good solution.

**Comment by daschuer** [comment]
Nice :-)  LGTM

**Pull request merged**